### PR TITLE
git: don't hardcode ssh to reduce rebuild amounts of openssh

### PR DESCRIPTION
**Pull request by SuperSandro2000**
Requirement for #146702

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/version-management/git-and-tools/git/default.nix [hunk 1]
```diff
@@ -1,6 +1,7 @@
 { fetchurl, lib, stdenv, buildPackages
 , curl, openssl, zlib, expat, perlPackages, python3, gettext, cpio
 , gnugrep, gnused, gawk, coreutils # needed at runtime by git-filter-branch etc
+, withSshSupport ? true
```
Comment (primeos) [inline pkgs/applications/version-management/git-and-tools/git/default.nix:4]: `withSshSupport` is misleading. SSH is supported either way, this is just about using a fixed SSH binary by hard-coding the Nix store path.

**Review by primeos — CHANGES_REQUESTED** [review]
Not a permanent blocker / hard veto but instead of adding yet another Git package I'd rather think about not hard-coding the SSH binary by default:
- https://github.com/NixOS/nixpkgs/issues/160527#issuecomment-1045114855
- https://github.com/NixOS/nixpkgs/issues/1923#issuecomment-1045118292

A dedicated issue for discussing this would be nice.

**Comment by SuperSandro2000** [comment]
Lets use this PR for that. Removing openssh makes my goal so much easier.

**Review by primeos — COMMENTED** [review]
+1 for this.

The commit message should also mention that this fixes https://github.com/NixOS/nixpkgs/issues/160527 too and contain a few sentences to explain the motivation for this change (e.g., advantages and drawbacks) before we merge this.

File: pkgs/applications/version-management/git-and-tools/git/default.nix [hunk 2]
```diff
@@ -48,22 +48,14 @@ stdenv.mkDerivation {
...
-
   patches = [
     ./docbook2texi.patch
     ./git-sh-i18n.patch
-    ./ssh-path.patch
     ./git-send-email-honor-PATH.patch
     ./installCheck-path.patch
   ];
 
   postPatch = ''
```
Comment (kevincox) [inline pkgs/applications/version-management/git-and-tools/git/default.nix:58]: This breaks cloning ssh repos. I don't think we should do this. I think we should rethink the possible solutions and go though an RFC before breaking tons of problems to work around a problem in Nix.

**Review by kevincox — CHANGES_REQUESTED** [review]

File: pkgs/applications/version-management/git-and-tools/git/default.nix (hunk 2 above)
Comment (primeos) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:58]: @kevincox what does this break? Git works and if you have SSH installed cloning via SSH works too. IMO that's how it should be. If there are enough relevant use-cases we could also add a wrapped Git to `all-packages.nix`.

Comment (kevincox) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:58]: The key phrase is "if you have SSH installed". If I am using `gitFull` I expect to be able to use all of the features of Git. Not get errors about missing dependencies.

I think a wrapped package for `gitFull` would be acceptable. That way the rebuild is trivial without having a disfunctional git package.

Comment (primeos) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:58]: > The key phrase is "if you have SSH installed".

Yes, and my point is that this should be the intended behaviour (SSH is a completely optional run time dependency and there are multiple SSH client packages available (plus different versions and configurations)).

> I think a wrapped package for gitFull would be acceptable.

That sounds fair to me.

Comment (SuperSandro2000) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:60]: I made the patch conditional for now because I don't want to deal with the implications of wrapping the git binary.

File: pkgs/applications/version-management/git-and-tools/git/default.nix [hunk 3]
```diff
@@ -27,7 +27,7 @@ assert svnSupport -> perlSupport;
 let
   version = "2.35.1";
   svn = subversionClient.override { perlBindings = perlSupport; };
-
+  isFull = svnSupport && guiSupport && sendEmailSupport;
```
Comment (kevincox) [inline pkgs/applications/version-management/git-and-tools/git/default.nix:30]: Can we have a separate value here? This seems like a hack. Like why not `pythonSupport`?

I think the best option is to use the presence of the openssh package as the flag.

File: pkgs/applications/version-management/git-and-tools/git/default.nix (hunk 2 above)
Comment (kevincox) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:60]: Thanks. I think this is a good tradeoff.

File: pkgs/applications/version-management/git-and-tools/git/default.nix (hunk 3 above)
Comment (primeos) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:30]: > I think the best option is to use the presence of the openssh package as the flag.

I like that idea :) For `git` it can default to `null` and then `gitFull` can override it to the actual package/default (`openssh`).

File: pkgs/applications/version-management/git-and-tools/git/default.nix [hunk 4]
```diff
@@ -4,15 +4,15 @@
 , openssh, pcre2, bash
 , asciidoc, texinfo, xmlto, docbook2x, docbook_xsl, docbook_xml_dtd_45
 , libxslt, tcl, tk, makeWrapper, libiconv
-, svnSupport, subversionClient, perlLibs, smtpPerlLibs
+, svnSupport ? false, subversionClient, perlLibs, smtpPerlLibs
```
Comment (primeos) [inline pkgs/applications/version-management/git-and-tools/git/default.nix:7]: Btw: If we add the defaults here we should probably also remove them from `git` in `pkgs/top-level/all-packages.nix`:
```
    svnSupport = false;         # for git-svn support
    guiSupport = false;         # requires tcl/tk
    sendEmailSupport = false;   # requires plenty of perl libraries
```

And IMO this should go into a dedicated commit as it's unrelated.

File: pkgs/applications/version-management/git-and-tools/git/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:30]: Overwriting packages to null is an anti pattern which I am not going to add more off.
I can also the flags from the default and minimal git to count as full or a new flag. Either ways are better.

File: pkgs/applications/version-management/git-and-tools/git/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/applications/version-management/git-and-tools/git/default.nix]: Moving that to another PR.

File: pkgs/applications/version-management/git-and-tools/git/default.nix (hunk 3 above)
Comment (kevincox) [reply pkgs/applications/version-management/git-and-tools/git/default.nix]: If you don't like the null package pattern then a `withSsh`  sounds good.

**Comment by Mindavi** [comment]
I think it's worth mentioning this in the release notes as well, since there may be people that will not have openssh installed systemwide and be surprised by this.

Other than that, reducing the number of rebuilds required is a good goal and I like this change.

**Review by kevincox — APPROVED** [review]

File: pkgs/applications/version-management/git-and-tools/git/default.nix [hunk 5]
```diff
@@ -18,6 +18,7 @@
 , withLibsecret ? false
 , pkg-config, glib, libsecret
 , gzip # needed at runtime by gitweb.cgi
+, hardcodeSsh ? false
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:21]: Yeah, `sshSupport` or `withSsh` does not really fit because you can use ssh remotes just fine when this is false. If someone knows a better name let me know.

[the remark this answers is not in the document]
Comment (kevincox) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:21]: I would still call it `withSsh`. I think that can be taken as this package isn't alone enough for SSH support. I agree that it is stretching the usual meaning a bit though.

File: pkgs/applications/version-management/git-and-tools/git/default.nix (hunk 3 above)
Comment (primeos) [reply pkgs/applications/version-management/git-and-tools/git/default.nix]: > Overwriting packages to null is an anti pattern

This is a personal opinion, unless it's documented somewhere (with consensus).
It's not ideal but so is the whole package customization situation.

> If you don't like the null package pattern then a `withSsh` sounds good.

Also sounds good to me.

File: pkgs/applications/version-management/git-and-tools/git/default.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/applications/version-management/git-and-tools/git/default.nix:21]: Fine for me. Pushing after I have generated the doc and then merging.

**Review by kevincox — APPROVED** [review]
